test/datahike/test/gc_test.cljc

```diff
@@ -0,0 +1,141 @@
+(ns datahike.test.gc-test
+  (:require
+   #?(:cljs [cljs.test :as t :refer-macros [is deftest testing]]
+      :clj  [clojure.test :as t :refer [is deftest testing]])
+   [clojure.set :as set]
+   [superv.async :refer [<?? S]]
+   [datahike.api :as d]
+   [datahike.experimental.gc :refer [gc!]]
+   [datahike.experimental.versioning :refer [branch! delete-branch! merge!
+                                             branch-history]]
+   [datahike.index.persistent-set :refer [mark]]
+   [konserve.core :as k]
+   [datahike.test.core-test])
+  (:import [java.util Date]))
+
+(defn- count-store [db]
+  (count (k/keys (:store db) {:sync? true})))
+
+(def count-query '[:find (count ?e) .
+                   :where
+                   [?e :age _]])
+
+(def txs (vec (for [i (range 1000)] {:age i})))
+
+(def cfg {:store              {:backend :file}
+          :keep-history?      true
+          :schema-flexibility :write
+          :index              :datahike.index/persistent-set})
+
+(def schema [{:db/ident       :age
+              :db/cardinality :db.cardinality/one
+              :db/valueType   :db.type/long}])
+
+(deftest datahike-gc-test
+  (testing "Test core garbage collector functionality."
+    (let [cfg (assoc-in cfg [:store :path] "/tmp/dh-gc-test")
+          conn (do
+                 (d/delete-database cfg)
+                 (d/create-database cfg)
+                 (d/connect cfg))
+          ;; everything will fit into the root nodes of each index here
+          num-roots 6
+          fresh-count (+ num-roots 3) ;; :branches + :db + cid + roots
+          ]
+      (is (= 1 (count (mark (:eavt @conn)))))
+      (is (= fresh-count (count-store @conn)))
+      (d/transact conn schema)
+      (is (= 1 (count (mark (:eavt @conn)))))
+      (is (= (+ 1 fresh-count num-roots) (count-store @conn)))
+      ;; delete old db with roots
+      (is (= (+ 1 num-roots) (count (<?? S (gc! @conn (Date.))))))
+      (is (= fresh-count (count-store @conn)))
+
+      ;; try to run on dirty index
+      (is (thrown-msg? "Index needs to be properly flushed before marking."
+                       (mark (:eavt
+                              (:db-after
+                               (d/with @conn [{:db/id 100
+                                               :age   5}]))))))
+
+      ;; check that we can still read the data
+      (d/transact conn txs)
+      (<?? S (gc! @conn (Date.)))
+      (is (= 1000 (d/q count-query @(d/connect cfg)))))))
```

Comment:
Instead of the `testing` statement at the top it would be more helpful to have several `testing` statements throughout the test describing what is being tested so problems are better visible on the test report. Here, it looks to me like the comments could be largely turned into `testing` statements enclosing the statements that they are concerned with.
A top level testing statement is usually not needed when the name of the test is chosen well.

Reply:
Agreed and fixed.